src/VisualStudio/Core/Def/Implementation/Snippets/SnippetFunctions/AbstractSnippetFunctionArgumentValue.cs

```diff
@@ -0,0 +1,74 @@
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using VsTextSpan = Microsoft.VisualStudio.TextManager.Interop.TextSpan;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Snippets
+{
+    internal abstract class AbstractSnippetFunctionArgumentValue : AbstractSnippetFunction
+    {
+        private readonly string _parameterName;
+        private readonly SymbolKey _parameter;
+
+        protected AbstractSnippetFunctionArgumentValue(AbstractSnippetExpansionClient snippetExpansionClient, ITextBuffer subjectBuffer, string fieldName, string parameter)
+            : base(snippetExpansionClient, subjectBuffer)
+        {
+            _parameterName = fieldName;
+            _parameter = new SymbolKey(parameter);
+        }
+
+        protected abstract string? FallbackDefaultLiteral { get; }
+
+        protected override int GetDefaultValue(CancellationToken cancellationToken, out string? value, out int hasCurrentValue)
+        {
+            if (!TryGetDocument(out var document))
+            {
+                value = null;
+                hasCurrentValue = 0;
+                return VSConstants.E_FAIL;
+            }
+
+            var textSpan = new VsTextSpan[1];
+            if (snippetExpansionClient.ExpansionSession.GetFieldSpan(_parameterName, textSpan) != VSConstants.S_OK)
+            {
+                value = null;
+                hasCurrentValue = 0;
+                return VSConstants.E_FAIL;
+            }
+
+            var text = document.GetTextSynchronously(cancellationToken);
+            var position = text.Lines.GetPosition(new LinePosition(textSpan[0].iStartLine, textSpan[0].iStartIndex));
+            var compilation = document.Project.GetRequiredCompilationAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+
+            if (!snippetExpansionClient.Arguments.TryGetValue(_parameterName, out var previousValue))
+                previousValue = null;
+
+            if (_parameter.Resolve(compilation, cancellationToken: cancellationToken).GetAnySymbol() is IParameterSymbol parameter)
+            {
+                foreach (var provider in snippetExpansionClient.GetArgumentProviders(document.Project.Solution.Workspace))
+                {
+                    var context = new ArgumentContext(provider, document, position, parameter, previousValue, cancellationToken);
+                    provider.ProvideArgumentAsync(context).Wait(cancellationToken);
+                    if (context.DefaultValue is not null)
+                    {
+                        value = context.DefaultValue;
+                        hasCurrentValue = 1;
+                        return VSConstants.S_OK;
+                    }
+                }
+            }
+
+            value = FallbackDefaultLiteral;
+            hasCurrentValue = value is not null ? 1 : 0;
+            return value is not null ? VSConstants.S_OK : VSConstants.E_FAIL;
```

Comment:
1. can you doc this method.  
2. my preference when mixing async/sync is to extract out an entirely async mehtod, then have a single point where the sync method calls the async one and blocks on it.

Reply:
⏱️ Extracted and documented asynchronous helper